staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,328 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+func bufferFromEvents(events []*watchCacheEvent) *watchCacheIntervalBuffer {
+	wcib := &watchCacheIntervalBuffer{}
+	wcib.buffer = make([]*watchCacheEvent, len(events))
+	copy(wcib.buffer, events)
+	wcib.endIndex = len(events)
+	return wcib
+}
+
+func TestIntervalBufferIsFull(t *testing.T) {
+	cases := []struct {
+		endIndex int
+		expected bool
+	}{
+		{bufferSize - 1, false},
+		{bufferSize, true},
+		{bufferSize + 1, true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{0, 10, false},
+		{50, 50, true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	events := []*watchCacheEvent{
+		{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+		{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+		{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+	}
+	wcib := bufferFromEvents(events)
+
+	for _, e := range events {
+		event, ok := wcib.next()
+		if !ok {
+			t.Error("unexpected bool value indicating empty buffer")
+		}
+		if !reflect.DeepEqual(event, e) {
+			t.Errorf("unexpected event, expected: %v, got %v", *event, *e)
+		}
+	}
+
+	event, ok := wcib.next()
+	if ok {
+		t.Error("unexpected bool value indicating buffer is not empty")
+	}
+	if event != nil {
+		t.Errorf("unexpected event received, expected: nil, got %v", *event)
+	}
+}
+
+func TestFillBuffer(t *testing.T) {
+	events := []*watchCacheEvent{
+		{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+		{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+		{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+	}
+	wci := intervalFromEvents(events)
+	wci.fillBuffer()
+	for i := 0; i < len(events); i++ {
+		if !reflect.DeepEqual(events[i], wci.buffer.buffer[i]) {
+			t.Errorf("expected: %v, got: %v", *events[i], *wci.buffer.buffer[i])
+		}
+	}
+}
+
+func TestCacheIntervalFromWatchCache(t *testing.T) {
```

Comment:
This is a good text, but I would rather treat it as a test of Next than then cacheInterval from watchcache.